proto/src/serializers/part_set_header_total.rs

```diff
@@ -0,0 +1,51 @@
+//! Serialize and deserialize part_set_header.total (from string or u32), (into u32 in
+//! part_set_header.total).
+//!
+//! The deserializer is created for backwards compatibility: `total` was changed from a
+//! string-quoted integer value into an integer value without quotes in Tendermint Core v0.34.0.
+//! This deserializer allows backwards-compatibility by deserializing both ways.
+//! See also: https://github.com/informalsystems/tendermint-rs/issues/679
+use serde::{de::Error, de::Visitor, Deserializer, Serialize, Serializer};
+use std::convert::TryFrom;
+use std::fmt::Formatter;
+
+struct PartSetHeaderTotalStringOrU32;
```

Comment:
Do we really need to define a type + a visitor?

I wonder if this can be expressed instead with just the `serialize` and `deserialize` functions, and invoked via a  `#[serde(with = "crate::serializers::part_set_header_total")]` attribute?

Reply:
Instead of a visitor, you can use the `deserialize` method provided by the `Deserialize` trait: https://docs.rs/serde/1.0.117/serde/trait.Deserialize.html

eg. to deserialize a `u64`:

```rust
u64::deserialize(deserializer)
```

Reply:
Yes, that was the original solution. Unfortunately, the deserializer in the above example is not clonable and I need to be able to derive if the incoming input is an integer or a string.

Here's the outline of the original solution:

```
let try_string = String::deserialize(deserializer.clone());
if try_string.is_err() {
  u32::deserialize(deserializer)?
} else {
  try_string.unwrap()
}
```

The deserialize function consumes the deserializer.

So, you need to be able to figure out if the incoming data is an integer or a String _before_ trying to deserialize it. Short of writing our own deserializer, the Visitor seemed the right solution. The "either string or struct" example in the Serde documentation also uses this solution: https://serde.rs/string-or-struct.html

Reply:
Right, makes sense! That's an unfortunate limitation/design decision of `serde`. Thanks for pointing me to the serde doc, had completely forgotten about this example,  my bad.

All good then :)

Reply:
No worries, I'm grateful that you point out possible optimizations!